File: 189-rotate-array/rotate-array.java
class Solution {
    public static void swap(int arr [], int i, int j){
        while(i <= j){
            int tm = arr[i];
            arr[i] = arr[j];
            arr[j] = tm;
            i++;
            j--;
        }
    }
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;
        swap(nums, 0, n-k-1);
        swap(nums, n-k, n-1);
        swap(nums, 0, n-1);
    }
}

File: 189-rotate-array/rotate-array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
    if(nums.length < 2) return;
    k = k % nums.length;
    let i = 0;
    let j = nums.length - 1;
    let l = j-k;
    while(i <= l){
        let temp = nums[i];
        nums[i] = nums[l];
        nums[l] = temp;
        i++;
        l--;
    }
    l = j-k+1;
    while(l <= j){
        let temp = nums[l];
        nums[l] = nums[j];
        nums[j] = temp;
        l++;
        j--;
    }
    i = 0;
    j = nums.length - 1;
    while(i <= j){
        let temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
        i++;
        j--;
    }
};